### PR TITLE
Add Hyena Operator and GPT2 Hyena

### PR DESCRIPTION
**Pull request by garymm**
HyenaOperator is an alternative to Attention.
Gpt2Hyena is like Gpt2, but using HyenaOperator instead of attention.

Fixes: #907

### CODE REVIEW

File: src/levanter/utils/activation.py
Comment (garymm) [inline src/levanter/utils/activation.py:1]: If you like this I can open a separate PR to use it throughout the repo and then rebase this PR on that.

File: src/levanter/models/gpt2_hyena.py
Comment (garymm) [inline src/levanter/models/gpt2_hyena.py:1]: I could potentially refactor Gpt2 to allow for less code duplication, but then it would be more abstract and less self-contained. LMK what you prefer.

File: src/levanter/models/hyena.py [hunk 1]
```diff
@@ -0,0 +1,540 @@
+"""An implementation of the Hyena operator.
+
+Paper: [Hyena Hierarchy: Towards Larger Convolutional Language Models](https://arxiv.org/abs/2302.10866)
+Official implementation in PyTorch:
+https://github.com/HazyResearch/safari/blob/541902aca88cb11af4d816ac762f3303e4ff8eea/src/models/sequence/hyena.py
+
+Current diffences from the official impl:
+- We don't support inner_factor.
+- We don't support post_order_ffn.
+- We don't support num_heads (the PyTorch impl's support for multiple heads seems incomplete, or at least
```
Comment (garymm) [inline src/levanter/models/hyena.py:10]: I could be wrong about this. LMK if you can tell if their code actually supports multiple heads, and if so please explain it to me and I will try to port it over.

**Review by garymm — COMMENTED** [review]
I have run the nano configuration which basically just ensures that it doesn't crash. I haven't verified if it actually learns. I will work on that next, but thought it might be valuable to put this up for review now.

**Comment by garymm** [comment]
Currently running the following on 2xH100, will let it run over night:

```yaml
data:
  id: dlwh/wikitext_103_detokenized

model:
  type: gpt2_hyena
  hidden_dim: 512
  num_layers: 8
  seq_len: 1024


trainer:
  tracker:
    tags: [ "wikitext", "gpt2_hyena"]

  mp: p=f32,c=bfloat16
  model_axis_size: 1
  checkpointer:
    keep:
      - every: 50
    save_interval: 5m

  per_device_parallelism: -1
  train_batch_size: 64

optimizer:
  learning_rate: 6E-4
  weight_decay: 0.1
  min_lr_ratio: 0.1
```

**Review by dlwh — COMMENTED** [review]
will look more later. a bit overwhelmd

File: src/levanter/models/gpt2_hyena.py
Comment (dlwh) [reply src/levanter/models/gpt2_hyena.py:1]: i think the copy paste is better probably.

File: src/levanter/models/hyena.py (hunk 1 above)
Comment (dlwh) [reply src/levanter/models/hyena.py:10]: doesn't it support num_heads:

https://github.com/HazyResearch/safari/blob/541902aca88cb11af4d816ac762f3303e4ff8eea/src/models/sequence/hyena.py#L317-L321

Comment (dlwh) [reply src/levanter/models/hyena.py:10]: oh right you said you don't really speak einsum. 

In their code:

```
rearrange(uc, 'b (ho v) (z l) -> b ho v z l', 
            z=self.num_blocks, 
            ho=self.num_heads, 
            v=self.head_dim * (self.order + 1)
```

amounts to:

```
b, d, seqlen = uc.shape
ho = self.num_blocks
v = self.head_dim * (self.order + 1)
z = self.num_blocks
l = seqlen // z
uc.reshape((b, ho, v, z, l)
```

so it reshapes the embedding dim into (heads, head size)

File: src/levanter/models/hyena.py [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+  I don't understand it).
+- We more efficiently support multiple blocks. I believe the PyTorch version is inefficient: It uses
+  the full sequence length in the PositionalEmbedding.
```
Comment (dlwh) [inline src/levanter/models/hyena.py:13]: but shouldn't it?

File: src/levanter/models/hyena.py [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+            Embed: Axis of positional embedding
+            key: Optional random key (not used)
+        """
+        seq_len = PosPerBlock.size
+
+        # Ensure emb_dim is valid
+        if Embed.size <= 1:
+            raise ValueError("emb_dim must be greater than 1")
+
+        # The time embedding fed to the filteres is normalized so that t_f = 1
```
Comment (dlwh) [inline src/levanter/models/hyena.py:105]:
```suggestion
        # The time embedding fed to the filters is normalized so that t_f = 1
```

File: src/levanter/models/hyena.py [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+        # Calculate number of frequency bands
+        bands = (Embed.size - 1) // 2
+        Band = Axis("band", bands)
+
+        # To compute the right embeddings we use the "proper" linspace
+        t_rescaled = hax.linspace(PosPerBlock, start=0, stop=seq_len - 1)
+        w = 2 * jnp.pi * t_rescaled / seq_len
+
+        f = hax.linspace(Band, start=1e-4, stop=bands - 1)
+        z_complex = hax.exp(-1j * f.broadcast_axis(PosPerBlock) * w.broadcast_axis(Band))
```
Comment (dlwh) [inline src/levanter/models/hyena.py:117]: i'm not checking this math so I'm assuming this is fine

**Review by garymm — COMMENTED** [review]
Thanks for the helpful review!

File: src/levanter/models/hyena.py [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+
+        return x
+
+
+class MLPTrainableActivation(eqx.Module):
+    """
+    hax.nn.MLP with a trainable activation function.
+    """
+
+    activation: Callable = eqx.field(static=False)  # this is the main difference from hax.nn.MLP
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/models/hyena.py:215]: yep you're definitely right!

[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/models/hyena.py:215]: maybe add to the docstirng that it's a shared activation?

File: src/levanter/models/hyena.py [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+        l_filter = min(input_length, Pos.size)
+
+        # Input projection from [Embed] to [(order+1) * Embed]
+        u = self.in_proj(u, key=key_in_proj)
+
+        # trying to keep the variable names from the official impl.
+        # I think uc stands for "u convolved".
+        uc = self.short_filter(u).rename({"out_channels": EmbedOrderPlus1})
+        # Note: after the short filter, uc."position" axis has slightly larger size due to padding.
+        # Hence we use Pos.name rather than Pos to slice, otherwise haliax complains the axis sizes
```
Comment (dlwh) [inline src/levanter/models/hyena.py:512]: what do you think about this behavior? I think I don't like it anymore and am thinking of changing it

File: src/levanter/utils/activation.py
Comment (dlwh) [reply src/levanter/utils/activation.py:1]: yeah i like! We can replace the ACT2FN map. (I do want lower case to still work I guess, but it's not a big deal)

**Review by dlwh — APPROVED** [review]
lgtm!

**Comment by dlwh** [comment]
i have a fix for test failures if you merge main
